### PR TITLE
Sided block opacity.

### PR DESCRIPTION
**Pull request by nevercast**
Implement isBlockSolid and isBlockSideSolid, as well as isSolid and isSideSolid on BlockType.
To be used with Torches, Levers and other blocks that are side surface dependent. Will add more code regarding this functionality later when I need it.

Josh Lloyd j.nevercast@gmail.com

### CODE REVIEW

**Comment by turt2live** [comment]
@nevercast Mind publishing your testing materials for this?

**Comment by nevercast** [comment]
Ooh! Professional. Uhh, how would I go about creation some 'testing material' ? :)

**Comment by Aaron1011** [comment]
@nevercast: Some kind of plugin that uses the functions you added - maybe logs the result, or displays it in the chat.

**Comment by nevercast** [comment]
They are only really meant to be internal functions used by placement logic of blocks, I've already started implementing it for everything from slabs to signs to torches. But I'll see what I can do.

File: src/main/java/net/glowstone/GlowWorld.java [hunk 1]
```diff
@@ -729,6 +733,24 @@ public boolean generateTree(Location loc, TreeType type, BlockChangeDelegate del
...
+        BlockType type = ItemTable.instance().getBlock(block.getTypeId());
+        return type.isSideSolid(block, face);
+    }
+
+    public boolean isBlockSolid(GlowBlock block) {
+        BlockType type = ItemTable.instance().getBlock(block.getTypeId());
+        return type.isSolid(block);
+    }
+
+    public boolean isBlockSolidOnTop(GlowBlock block) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/GlowWorld.java]: This (and the other `___OnTop` methods) should be removed - they are kinda useless seeing as they just delegate to other methods.

Comment (nevercast) [inline src/main/java/net/glowstone/GlowWorld.java]: That was their intention, it made code elsewhere in the project look more tidy by reducing arguments and increasing clarity. But removing these is no issue.

**Comment by gdude2002** [comment]
This has Checkstyle errors: http://bamboo.gserv.me/browse/GLOW-SRV-PR272-47

**Comment by turt2live** [comment]
@nevercast Still waiting on those changes to be made before I can test this further

**Comment by turt2live** [comment]
@nevercast Your PR does not compile! Here's some information about the compile testing:
- Glowkit PR found: false
- Glowkit PR parsed: _NONE_
- Glowkit compiled: false

If your PR requires Glowkit, please ensure that an accurate link to the Glowkit PR appears in the description of your PR. If the information above is incorrect, please contact me on IRC (`turt2live` in `#glowstone` on esper.net) for more information.

If your PR does not require Glowkit, please fix your PR.

Thank you!

`PR_BOT_CODE:3` (you can ignore this)

**Comment by dequis** [comment]
This PR fails to build due to recent glowkit changes. Merging fixes that, but then there are checkstyle errors.

Fixed version here: https://github.com/dequis/Glowstone/compare/pr-272

Also submitted PR to @nevercast

**Comment by nevercast** [comment]
I for some reason have not been getting notifications of these comments.
The thing compiled back when I opened the PR, @dequis thanks for creating that PR I'll merge that in to my repo.

I'll merge that in and look at what else needs fixing this evening.

**Comment by nevercast** [comment]
Merged and build passing, can a checkstyle be run?
I'll do some amending tonight and squish commits if needed.

Is a unit testing framework in use on Glowstone? I've recently become familiar with BDD style tests

File: src/main/java/net/glowstone/GlowWorld.java (hunk 1 above)
Comment (turt2live) [inline src/main/java/net/glowstone/GlowWorld.java]: These methods are still present. Please remove them

**Comment by nevercast** [comment]
Tag added: **Needs Changes**
- [x] Remove useless helpers (Like the top ones per @turt2live 's suggestion below)
- [ ] Write a test for this

**Comment by dequis** [comment]
Checkstyle and unit testing is done by gradle on every build now, so those are required to build. I already included a commit fixing the checkstyle issues.

Glowstone uses junit, look at ./src/test/java/net/glowstone. And yeah I guess unit testing would be the sanest way to prove that a PR like this one works.

**Comment by turt2live** [comment]
@nevercast Don't really have to remove all the helpers, just the ones that are more or less a waste of time (like the `top` ones, you can just pass a block face to another helper)

**Comment by nevercast** [comment]
No problem.

**Comment by gdude2002** [comment]
Passing: http://bamboo.gserv.me/browse/GLOW-SRV-PR272-91

**Comment by nevercast** [comment]
Thanks @gdude2002 
I've been swamped at work so wont get on to this until the weekend.

**Comment by turt2live** [comment]
@nevercast the weekend passed :(

**Comment by nevercast** [comment]
@turt2live Sorry man :( I don't get the time I'd like for hobbies anymore.
I see you've assigned it to yourself. I'm subscribed here so will keep an ear out.

**Comment by dequis** [comment]
Self-assigned to turt2live == he will take care of review.

But noted, added this to the list of PRs for adoption

**Comment by turt2live** [comment]
Calling upon @dequis to complete/work on this PR. Unless @nevercast becomes undead.

**Comment by dequis** [comment]
Yay, I'm officially the guy who takes care of the corpses.

**Comment by dequis** [comment]
Okay merged the trivial conflict here (imports) and added a commit that removes the useless helpers:

https://github.com/dequis/Glowstone/compare/pr-272

I found it hard to test this damn thing. Unit tests are NOPE - the interesting parts of this PR require GlowWorld, and our unit tests use a shitty server shim that has no world.

Then I tried to make a [test plugin](https://gist.github.com/dequis/b0872cfca39d0bee7f26) that checks the PlayerInteractEvent result, and tried placing some ladders/vines/levers against slabs/stairs/glass/etc. The behavior did not change at all with or without this PR. I have no idea what behavior should be changed by this. Everything worked correctly, but that might be the client being too nice.

**Comment by turt2live** [comment]
@nevercast To be honest I have no idea what this PR aims to achieve, let alone how to test is perfectly. As @dequis mentioned, the standard approach to testing this unknown PR yield no changes, so I assume it is some kind of performance fix or something along those lines.

You do make mention to implementing this PR in the future if the PR is pulled, but I personally do not see a clear implementation. Do you have a sample branch we can look at to see how this can be used? As it stands right now the PR is in a confusing state. I've put it to "needs changes" for pending a response to this message.

Thanks!

**Comment by nevercast** [comment]
@turt2live, @dequis  
I was working on redstone support for Glowstone, a nice clean implementation that followed all the quirks of vanilla as I'm an avid redstone builder. I was using this PR for torches and redstone place logic.

It's purpose was to expose an API for blocks to say if a side was solid for an attaching block to be placed against it.
An example of attaching blocks are Levers, buttons, signs, item frames, pictures (Though they spawn entities, similar logic applies), torches, redstone etc. This code was so than when each of these blocks were implemented you could handle placement via the API. Also when a solid block was created it could override a handler to specify if a torch could be stuck to it's side for example.

The isSolid/is3D don't completely apply here as some blocks may be Solid but you cannot place torches on them.

Saying that I've abandoned it would be an understatement. I'll setup Glowstone right this second and see if I can restore this to life with tests, or discard it entirely.

**Comment by nevercast** [comment]
Currently it's difficult to test this in Vanilla because the expected actions aren't reflected in 1.8, there is possibly regression in the latest Minecraft.

Place a lever on top of some inverted stairs and power the stairs with a redstone torch under it. (Likely a block update that is causing it)

The expected action is that the lever cannot power the stairs as it's not solid, but because the topside (bottom if inverted) is solid, you can still place and operate a lever there due to it powering neighbouring blocks.

As the lever instead drops as an item, there is an issue in Vanilla, meaning around placement logic there could be more than one issue, giving me no accurate implementation to model the placement logic against.

**Comment by dequis** [comment]
Oh hey, thanks for respawning!

There have been some other PRs that work on redstone support, such as #436 and #439 (both continuations of the original, #215). If you think this would be easier to test with those, then we can put this on one hold for testing.

Regarding tests: a bukkit plugin that checks whatever needs to be checked on PlayerInteractEvent is probably the sanest way to do it. Like I said before, I'm not a fan of our unit testing, so bukkit plugins all the way.

**Comment by nevercast** [comment]
@dequis 

Sorry about the limbo, have been busy with IRL work and not had time to work on this, as for the other PRs I've looked through them and they seem overly complex for something so simple, redstone logic does not differ from normal block update logic that should exist in the game engine as just BlockManager. This should be able to send tick events to blocks that require them, such as repeaters, torches, comparators.... trees, wheat etc.

I'm not entirely sure why there is a need for an RSManager at all unless there is intention to make an interface out of it and provide the option of alternate redstone logic with plugins.

**Comment by nevercast** [comment]
Required for working redstone:
- [ ] Placement logic involving placing items on surfaces that allow it, not all blocks can support torches
- [ ] Block update logic, block removal and metadata updates should cause updates on neighbours
- [ ] Utilization of block update logic to cause 'circuits' to drop as item when parent block is broken
- [ ] Block updates are used to update redstone power level, the logic for this is stated below
- [ ] A single global list tracks each torch block and the amount of times it is toggled per tick, (burn out)
- [ ] Chunk tick handler that can cause a block update on a block at a later time, (scheduler)
- [ ] API for isPowered, isDirectlyPowered, canBePowered, isEmittingPower etc for redstone blocks

Only the redstone power transmission and the list that tracks torch burnout need to be implemented for working redstone, everything else should be implemented anyway due to dependencies from other parts of the server.

Redstone update is simple, when redstone dust gets updated it's new metadata value is the result of:
`clamp( isDirectlyPowered ? 15 : max( neighboringRedstone.power ) - 1, 0, 15)`
Explained: If directly powered, 15, otherwise it is the max adjacent redstone power, minus 1 , clamped between 0 and 15. The metadata is then updated with this value.

The metadata setter should cause a block update when the metadata is different to the current value for correct operation. The API that handles changing of metadata should offer the option to not send the changes back to the client, and not cause a block update, this helps for large updates like explosions, giving the option to commit the changes later.

So TL;DR. Redstone uses simple mechanics that should already be implemented for other uses

**Comment by nevercast** [comment]
For the record, this PR was going to take care of the first bullet point on that list.

**Comment by turt2live** [comment]
So if I understand this correctly, this PR simply makes it so you can't put something on a non-solid surface of a block (ie: the top of the stairs). But from my understanding, there is no item that can be placed there anyways, so what does this PR accomplish? 

I'm having trouble understanding what this PR is for because it has little to no implementation and a lot of code. There is no apparent use case or apparent required change to make this worthwhile. I honestly simply do not see what purpose this serves in Glowstone at the moment. There is a lot of mention about redstone, but all of my understanding of redstone still renders this not required.

**Comment by nevercast** [comment]
The reason was to make the server-side assertive of block placement rules, as clients could place a torch anywhere they wanted to. Also dropping of torches and similar was to be implemented through this. I'll put this PR in limbo until it's use case is obvious or there are other implementations.
